shell/platform/embedder/embedder_semantics_update.cc

```diff
@@ -175,4 +193,61 @@ void EmbedderSemanticsUpdate2::AddAction(
   });
 }
 
+std::pair<size_t, const FlutterStringAttribute**>
+EmbedderSemanticsUpdate2::CreateStringAttributes(
+    const StringAttributes& attributes) {
+  // Minimize allocations if attributes are empty.
+  if (attributes.empty()) {
+    return std::make_pair(0, nullptr);
+  }
+
+  auto result = std::make_unique<std::vector<const FlutterStringAttribute*>>();
+  result->reserve(attributes.size());
+
+  for (const auto& attribute : attributes) {
+    auto embedder_attribute = std::make_unique<FlutterStringAttribute>();
+    embedder_attribute->struct_size = sizeof(FlutterStringAttribute);
+    embedder_attribute->start = attribute->start;
+    embedder_attribute->end = attribute->end;
+
+    switch (attribute->type) {
+      case StringAttributeType::kLocale: {
+        std::shared_ptr<flutter::LocaleStringAttribute> locale_attribute =
+            std::static_pointer_cast<flutter::LocaleStringAttribute>(attribute);
```

Comment:
Alright - looking at [string_attribute.h:15-16](https://github.com/flutter/engine/blob/714aa98750c0a14ca4eebaab6327817ba368d507/lib/ui/semantics/string_attribute.h#L15-L16) and [string_attribute.h:49-51](https://github.com/flutter/engine/blob/714aa98750c0a14ca4eebaab6327817ba368d507/lib/ui/semantics/string_attribute.h#L49-51), I see why where the `shared_ptr`s are coming from. I haven't looked into that code do see why they're necessary, but I'll take it on faith that they are :)